File: day01-js+go/day1.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func abs(x int) int {
	return x < 0 {
		return -x
	}
	return x
}

func main() {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	aValues, bValues := make([]int, len(lines)), make([]int, len(lines))

	for i, line := range lines {
		parts := strings.Fields(line)
		aValues[i], _ = strconv.Atoi(parts[0])
		bValues[i], _ = strconv.Atoi(parts[1])
	}

	sort.Ints(aValues)
	sort.Ints(bValues)

	task1Result := 0
	for i := range aValues {
		task1Result += abs(aValues[i] - bValues[i])
	}

	bAppearances := map[int]int{}
	for _, b := range bValues {
		bAppearances[b]++
	}

	task2Result := 0
	for _, a := range aValues {
		task2Result += a * bAppearances[a]
	}

	fmt.Printf("Score for task 1 is %d\n", task1Result)
	fmt.Printf("Score for task 2 is %d\n", task2Result)
}

File: day01-js+go/day1.mjs
import { open } from 'node:fs/promises';

const file = await open('input.txt');

const aValues = [];
const bValues = [];

for await (const line of file.readLines()) {
    const [aValue, bValue] = line.split('   ').map(Number);
    aValues.push(aValue);
    bValues.push(bValue);
}

aValues.sort((a1, a2) => a1 - a2);
bValues.sort((b1, b2) => b1 - b2);
const distances = aValues.map((a, i) => Math.abs(a - bValues[i]));
const task1Result = distances.reduce((acc, el) => acc + el);

console.log(`Score for task 1 is ${task1Result}`);

const bAppearances = bValues.reduce((acc, val) => {
    acc[val] ??= 0;
    acc[val]++;
    return acc;
}, {});
const task2Result = aValues.reduce(
    (acc, val) => acc + val * (bAppearances[val] ?? 0),
    0
);

console.log(`Score for task 2 is ${task2Result}`);
